File: results/0.25/NITF/nitf-gpt-4-turbo.c
#include <hammer/hammer.h>
#include <stdio.h>
#include <stdlib.h>

// Define parsers for various fields
HParser *parse_fhdr() {
    return h_sequence(h_ch_range('0', '9'), h_ch_range('0', '9'), h_ch_range('0', '9'), h_ch_range('0', '9'), h_ch_range('0', '9'), h_ch_range('0', '9'), NULL);
}

HParser *parse_clevel() {
    return h_int_range(h_uint8(), 1, 99);
}

HParser *parse_stype() {
    return h_token("BF01", 4);
}

HParser *parse_ostaid() {
    return h_token("ABCDEFGHIJ", 10);
}

HParser *parse_fdt() {
    return h_sequence(h_ch_range('0', '9'), h_ch_range('0', '9'), h_ch_range('0', '9'), h_ch_range('0', '9'), h_ch_range('0', '9'), h_ch_range('0', '9'), h_ch_range('0', '9'), h_ch_range('0', '9'), h_ch_range('0', '9'), h_ch_range('0', '9'), h_ch_range('0', '9'), h_ch_range('0', '9'), h_ch_range('0', '9'), h_ch_range('0', '9'), NULL);
}

HParser *parse_fsclas() {
    return h_choice(h_ch('U'), h_ch('R'), h_ch('C'), h_ch('S'), h_ch('T'), NULL);
}

HParser *parse_field_generic(int length) {
    return h_repeat_n(h_uint8(), length);
}

HParser *parse_nitf_header() {
    return h_sequence(
        parse_fhdr(),
        parse_clevel(),
        parse_stype(),
        parse_ostaid(),
        parse_fdt(),
        parse_fsclas(),
        parse_field_generic(2), // FSCLSY
        parse_field_generic(11), // FSCODE
        parse_field_generic(2), // FSCTLH
        parse_field_generic(20), // FSREL
        parse_field_generic(2), // FSDCTP
        parse_field_generic(8), // FSDCDT
        parse_field_generic(4), // FSDCXM
        parse_field_generic(1), // FSDG
        parse_field_generic(8), // FSDGDT
        parse_field_generic(43), // FSCLTX
        parse_field_generic(1), // FSCATP
        parse_field_generic(40), // FSCAUT
        parse_field_generic(1), // FSCRSN
        parse_field_generic(8), // FSSRDT
        parse_field_generic(15), // FSCTLN
        parse_field_generic(5), // FSCOP
        parse_field_generic(5), // FSCPYS
        parse_field_generic(1), // ENCRYP
        parse_field_generic(3), // FBKGC
        parse_field_generic(24), // ONAME
        parse_field_generic(18), // OPHONE
        NULL
    );
}

// Main function to parse a NITF file
int main(int argc, char **argv) {
    if (argc != 2) {
        fprintf(stderr, "Usage: %s <nitf_file>\n", argv[0]);
        return 1;
    }

    FILE *fp = fopen(argv[1], "rb");
    if (!fp) {
        perror("File opening failed");
        return EXIT_FAILURE;
    }

    fseek(fp, 0, SEEK_END);
    size_t filesize = ftell(fp);
    fseek(fp, 0, SEEK_SET);

    uint8_t *buffer = malloc(filesize);
    if (!buffer) {
        perror("Memory allocation failed");
        fclose(fp);
        return EXIT_FAILURE;
    }

    fread(buffer, 1, filesize, fp);
    fclose(fp);

    HParser *nitf_parser = parse_nitf_header();
    HParseResult *result = h_parse(nitf_parser, buffer, filesize);

    if (result) {
        printf("NITF file parsed successfully.\n");
    } else {
        printf("Failed to parse NITF file.\n");
    }

    h_parse_result_free(result);
    h_parser_unref(nitf_parser);
    free(buffer);

    return 0;
}

File: results/0.25/NITF/nitf-gpt-4-turbo.rs
use nom::{
    bytes::complete::take,
    combinator::{map_res, all_consuming},
    sequence::tuple,
    IResult,
};
use std::fs::File;
use std::io::{self, Read};
use std::env;

#[derive(Debug)]
struct FileHeader {
    fhdr: String,
    fver: String,
    clevel: String,
    stype: String,
    osta_id: String,
    fdt: String,
    ftitle: String,
    fsclas: String,
    fsclsy: String,
    fscode: String,
    fsctlh: String,
    fsrel: String,
    fsdctp: String,
    fsdcdt: String,
    fsdcxm: String,
    fsdgor: String,
    fscaut: String,
    fsctln: String,
    fscop: String,
    fscpys: String,
    encrypt: String,
    fbkgc: String,
    oname: String,
    ophone: String,
}

fn parse_fixed_length(input: &[u8], length: usize) -> IResult<&[u8], String> {
    map_res(take(length), std::str::from_utf8)(input).map(|(next_input, result)| (next_input, result.to_string()))
}

fn parse_file_header(input: &[u8]) -> IResult<&[u8], FileHeader> {
    let parser = tuple((
        parse_fixed_length(9),
        parse_fixed_length(2),
        parse_fixed_length(2),
        parse_fixed_length(4),
        parse_fixed_length(10),
        parse_fixed_length(14),
        parse_fixed_length(80),
        parse_fixed_length(1),
        parse_fixed_length(2),
        parse_fixed_length(11),
        parse_fixed_length(2),
        parse_fixed_length(20),
        parse_fixed_length(2),
        parse_fixed_length(8),
        parse_fixed_length(4),
        parse_fixed_length(1),
        parse_fixed_length(40),
        parse_fixed_length(15),
        parse_fixed_length(5),
        parse_fixed_length(5),
        parse_fixed_length(1),
        parse_fixed_length(3),
        parse_fixed_length(24),
        parse_fixed_length(18),
    ));
    let (input, (fhdr, fver, clevel, stype, osta_id, fdt, ftitle, fsclas, fsclsy, fscode, fsctlh, fsrel, fsdctp, fsdcdt, fsdcxm, fsdgor, fscaut, fsctln, fscop, fscpys, encrypt, fbkgc, oname, ophone)) = all_consuming(parser)(input)?;

    Ok((input, FileHeader {
        fhdr, fver, clevel, stype, osta_id, fdt, ftitle, fsclas, fsclsy, fscode, fsctlh, fsrel, fsdctp, fsdcdt, fsdcxm, fsdgor, fscaut, fsctln, fscop, fscpys, encrypt, fbkgc, oname, ophone,
    }))
}

fn main() -> io::Result<()> {
    let args: Vec<String> = env::args().collect();
    if args.len() != 2 {
        println!("Usage: {} <filename>", args[0]);
        return Ok(());
    }

    let filename = &args[1];
    let mut file = File::open(filename)?;
    let mut buffer = Vec::new();
    file.read_to_end(&mut buffer)?;

    match parse_file_header(&buffer) {
        Ok((_, header)) => println!("{:?}", header),
        Err(e) => println!("Failed to parse NITF file header: {:?}", e),
    }

    Ok(())
}